### ISSUE TITLE
fix "IPv4: Attempt to release TCP socket in state 1 "... on shutdown

### ISSUE THREAD

**Issue by pabeni** [multipath-tcp/mptcp_net-next#9]
running the following syzkaller:

```
# {Threaded:false Collide:false Repeat:false RepeatTimes:0 Procs:1 Sandbox: Fault:false FaultCall:-1 FaultNth:0 Leak:false NetInjection:false NetDevices:false NetReset:false Cgroups:false BinfmtMisc:false CloseFDs:false KCSAN:false DevlinkPCI:false UseTmpDir:false HandleSegv:false Repro:false Trace:false}
r0 = socket$inet_mptcp(0x2, 0x1, 0x106)
r1 = socket$inet_mptcp(0x2, 0x1, 0x106)
bind$inet(r1, &(0x7f00000013c0)={0x2, 0x4e20, @multicast2}, 0x10)
connect$inet(r1, &(0x7f0000000040)={0x2, 0x0, @loopback}, 0x10)
listen(r1, 0x3)
connect$inet(r0, &(0x7f0000000040)={0x2, 0x4e20, @loopback}, 0x4d)
sendmsg$inet(r0, &(0x7f0000000280)={0x0, 0x0, &(0x7f0000000000)=[{&(0x7f0000000080)="ff", 0x20000081}], 0x1}, 0x0)

```
beyond triggering issues/3, it blocks - as expected. Explicitly killing the process yeld the following warning:

"IPv4: Attempt to release TCP socket in state 1"

The call trace to the above printk is the following one:

            7fff82f81a4a printk ([kernel.kallsyms])
            7fff8441f637 __sk_destruct ([kernel.kallsyms])
            7fff84aaecbf subflow_ulp_release ([kernel.kallsyms])
            7fff847747ae tcp_cleanup_ulp ([kernel.kallsyms])
            7fff8474c812 tcp_v4_destroy_sock ([kernel.kallsyms])
            7fff846c383a inet_csk_destroy_sock ([kernel.kallsyms])
            7fff846c42ad inet_csk_listen_stop ([kernel.kallsyms])
            7fff846e3f8c tcp_close ([kernel.kallsyms])
            7fff847c2ea6 inet_release ([kernel.kallsyms])
            7fff84408605 __sock_release ([kernel.kallsyms])
            7fff84a9fda3 mptcp_close ([kernel.kallsyms])
            7fff847c2ea6 inet_release ([kernel.kallsyms])
            7fff84408515 __sock_release ([kernel.kallsyms])
            7fff844086e4 sock_close ([kernel.kallsyms])
            7fff83591224 __fput ([kernel.kallsyms])
            7fff82e5b789 task_work_run ([kernel.kallsyms])
            7fff82df7c3a do_exit ([kernel.kallsyms])
            7fff82df9ded do_group_exit ([kernel.kallsyms])
            7fff82e24a5e get_signal ([kernel.kallsyms])
            7fff82c6777d do_signal ([kernel.kallsyms])
            7fff82c0c51f exit_to_usermode_loop ([kernel.kallsyms])
            7fff82c0ee88 do_syscall_64 ([kernel.kallsyms])
            7fff84c05091 entry_SYSCALL_64_after_hwframe ([kernel.kallsyms])

**Comment by pabeni** [comment]
The issue with stack trace reported above is fixed by upstream commit:

df1036da90108b1a9969721beab34f4c76228bcc

But I'm still hitting the warn message with a different stack trace:

```
[  473.935207] IPv4: Attempt to release TCP socket in state 1 00000000e32b8b7e
[  473.939298] CPU: 1 PID: 6470 Comm: repro Not tainted 5.7.0-rc1.mptcp_3406c7fa5775+ #85
[  473.941191] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS 1.12.0-2.fc30 04/01/2014
[  473.965780] Call Trace:
[  473.967752]  <IRQ>
[  473.976341] dump_stack (/home/pabeni/devel/linux-mptcp/lib/dump_stack.c:120)
[  473.980188] __sk_destruct (/home/pabeni/devel/linux-mptcp/./include/linux/compiler.h:199 (discriminator 1) /home/pabeni/devel/linux-mptcp/net/core/sock.c:1698 (discriminator 1))
[  473.981238] subflow_ulp_release (/home/pabeni/devel/linux-mptcp/./include/net/sock.h:1778 /home/pabeni/devel/linux-mptcp/net/mptcp/subflow.c:1120)
[  473.986470] tcp_cleanup_ulp (/home/pabeni/devel/linux-mptcp/net/ipv4/tcp_ulp.c:124)
[  473.987506] tcp_v4_destroy_sock (/home/pabeni/devel/linux-mptcp/net/ipv4/tcp_ipv4.c:2164)
[  473.992193] inet_csk_destroy_sock (/home/pabeni/devel/linux-mptcp/net/ipv4/inet_connection_sock.c:878 (discriminator 9))
[  473.993215] tcp_reset (/home/pabeni/devel/linux-mptcp/./include/asm-generic/bitops/instrumented-non-atomic.h:110 /home/pabeni/devel/linux-mptcp/./include/net/sock.h:884 /home/pabeni/devel/linux-mptcp/net/ipv4/tcp_input.c:4137)
[  473.994024] tcp_validate_incoming (/home/pabeni/devel/linux-mptcp/net/ipv4/tcp_input.c:5470)
[  473.995184] tcp_rcv_state_process (/home/pabeni/devel/linux-mptcp/net/ipv4/tcp_input.c:6232)
[  474.006715] tcp_v4_do_rcv (/home/pabeni/devel/linux-mptcp/net/ipv4/tcp_ipv4.c:1643)
[  474.009066] tcp_v4_rcv (/home/pabeni/devel/linux-mptcp/./include/linux/spinlock.h:393 /home/pabeni/devel/linux-mptcp/net/ipv4/tcp_ipv4.c:2009)
[  474.019255] ip_protocol_deliver_rcu (/home/pabeni/devel/linux-mptcp/net/ipv4/ip_input.c:204 (discriminator 1))
[  474.020764] ip_local_deliver_finish (/home/pabeni/devel/linux-mptcp/./include/linux/rcupdate.h:651 /home/pabeni/devel/linux-mptcp/net/ipv4/ip_input.c:232)
[  474.022248] ip_local_deliver (/home/pabeni/devel/linux-mptcp/./include/linux/netfilter.h:307 /home/pabeni/devel/linux-mptcp/./include/linux/netfilter.h:301 /home/pabeni/devel/linux-mptcp/net/ipv4/ip_input.c:252)
[  474.031541] ip_rcv (/home/pabeni/devel/linux-mptcp/./include/linux/netfilter.h:307 /home/pabeni/devel/linux-mptcp/./include/linux/netfilter.h:301 /home/pabeni/devel/linux-mptcp/net/ipv4/ip_input.c:539)
[  474.045337] __netif_receive_skb_one_core (/home/pabeni/devel/linux-mptcp/net/core/dev.c:5188 (discriminator 4))
[  474.061073] process_backlog (/home/pabeni/devel/linux-mptcp/./include/linux/rcupdate.h:651 /home/pabeni/devel/linux-mptcp/net/core/dev.c:6135)
[  474.070095] net_rx_action (/home/pabeni/devel/linux-mptcp/net/core/dev.c:6572 /home/pabeni/devel/linux-mptcp/net/core/dev.c:6640)
[  474.088098] __do_softirq (/home/pabeni/devel/linux-mptcp/./arch/x86/include/asm/jump_label.h:25 /home/pabeni/devel/linux-mptcp/./include/linux/jump_label.h:200 /home/pabeni/devel/linux-mptcp/./include/trace/events/irq.h:142 /home/pabeni/devel/linux-mptcp/kernel/softirq.c:293)
[  474.092101] do_softirq_own_stack (/home/pabeni/devel/linux-mptcp/arch/x86/entry/entry_64.S:1083)
[  474.093820]  </IRQ>
[  474.100718] do_softirq.part.0 (/home/pabeni/devel/linux-mptcp/kernel/softirq.c:337)
[  474.104535] __local_bh_enable_ip (/home/pabeni/devel/linux-mptcp/./arch/x86/include/asm/preempt.h:26 /home/pabeni/devel/linux-mptcp/kernel/softirq.c:189)
[  474.110930] inet_csk_listen_stop (/home/pabeni/devel/linux-mptcp/./include/asm-generic/atomic-instrumented.h:199 /home/pabeni/devel/linux-mptcp/./include/linux/refcount.h:266 /home/pabeni/devel/linux-mptcp/./include/linux/refcount.h:294 /home/pabeni/devel/linux-mptcp/./include/net/sock.h:1777 /home/pabeni/devel/linux-mptcp/net/ipv4/inet_connection_sock.c:1034)
[  474.112767] tcp_close (/home/pabeni/devel/linux-mptcp/net/ipv4/tcp.c:2355)
[  474.113536] inet_release (/home/pabeni/devel/linux-mptcp/net/ipv4/af_inet.c:429)
[  474.119945] __sock_release (/home/pabeni/devel/linux-mptcp/net/socket.c:606)
[  474.127775] mptcp_close (/home/pabeni/devel/linux-mptcp/net/mptcp/protocol.c:1287)
[  474.139277] inet_release (/home/pabeni/devel/linux-mptcp/net/ipv4/af_inet.c:429)
[  474.142805] __sock_release (/home/pabeni/devel/linux-mptcp/net/socket.c:606)
[  474.146625] sock_close (/home/pabeni/devel/linux-mptcp/net/socket.c:1285)
[  474.148927] __fput (/home/pabeni/devel/linux-mptcp/fs/file_table.c:280)
[  474.152196] task_work_run (/home/pabeni/devel/linux-mptcp/kernel/task_work.c:125 (discriminator 1))
[  474.154412] exit_to_usermode_loop (/home/pabeni/devel/linux-mptcp/./include/linux/tracehook.h:188 /home/pabeni/devel/linux-mptcp/arch/x86/entry/common.c:165)
```
this looks related to issue#8

**Comment by pabeni** [comment]
The latter splat is really caused by issue#8, will be handled there. Closing this one as fixed by commit df1036da90108b1a9969721beab34f4c76228bcc
